adjust 2D model to match gait2392

antoinefalisse opened the pull request:
Hey, this PR is a first attempt to fix #129 

I replaced path points, optimal fiber lengths, and tendon slack lengths of the 2D model by those of gait2392.

I used:
- hamstrings (2D) <-> BFLH (gait2392)
- glut_max (2D) <-> glut_max2 (gait2392)
- vast (2D) <-> vas_int (gait2392)
- Iliopsoas (2D) <-> psoas (gait2392)
- gastroc (2D) <-> med_gas (gait2392)

I verified that:
- Both models were symmetrical.
- Both models had the same muscle-tendon parameters (except for Fmax).
- Both models had the same lengths and moment arms for a set of random poses (except for the ankle muscles, since the ankle definition is different in 3D vs 2D).

I wrote a MATLAB script (`compare_2D_vs_3D.m`) for those tests (see zip file). If you want to try out yourself, make sure you first run the muscle analyses for both models (the setup files are ready, just need `opensim-cmd run-tool SetupMA`).

[test2DModel.zip](https://github.com/opensim-org/opensim-models/files/4948913/test2DModel.zip)

Several muscle-tendon parameters were quite different between the original gait10dof18musc and gait2392 so it might be worth discussing why before changing everything. I read in the documentation that @aseth1 adjusted the model in the first place so I don't want to change something that had been done on purpose.

Note that there are a few other scaled versions of that 2D model, those should also be adjusted if we go forward with this PR. But I first wanted to have feedback on this.

Thanks in advance.

carmichaelong commented:
@antoinefalisse Thanks for these changes! I just looked through this and double checked the changes against gait2392 and it all looks good. I just saw some minor inconsistency with tab usage, so I may a minor commit on top (I hope you don't mind!). Let me know if there's anything else that might have changed since you made this PR, otherwise feel free to merge.

antoinefalisse commented:
Thanks @carmichaelong ! I guess we should also adjust the other models in the Gait10dof18musc folder for consistency. It looks like those models are scaled so we can probably use the scaled values from the scaled version of gait2392 -> [this model](https://github.com/opensim-org/opensim-models/blob/master/Models/Gait2392_Simbody/subject01_adjusted.osim). What do you think? I will take a closer look in the coming days to check if that makes sense.

carmichaelong commented:
Ah good point. I did a quick diff on the file you linked in the repo to the `Gait10dof18musc\subject01.osim` file. Looks like these two scaled files are related, as *almost* all of the skeletal properties are the same to within 6 or 7 decimal digits. The only thing that differed by a substantial amount was oddly the torso's `mass_center` z-component(?!)
![image](https://user-images.githubusercontent.com/7455790/94849209-e9c94380-03d9-11eb-86fe-a0c7c30eb6e9.png)

antoinefalisse commented:
@carmichaelong sorry for the delay. 

I think the torso mass center of [Gait10dof18musc/subject01.osim](https://github.com/opensim-org/opensim-models/blob/master/Models/Gait10dof18musc/subject01.osim#L1297) is the same as the one of [Gait2392_Simbody/subject01.osim](https://github.com/opensim-org/opensim-models/blob/master/Models/Gait2392_Simbody/subject01.osim#L2247). But there is a difference between the torso mass center of [Gait2392_Simbody/subject01.osim](https://github.com/opensim-org/opensim-models/blob/master/Models/Gait2392_Simbody/subject01.osim#L2247) and the one of [Gait2392_Simbody/subject01_adjusted.osim](https://github.com/opensim-org/opensim-models/blob/master/Models/Gait2392_Simbody/subject01_adjusted.osim#L2247) you referred to. I don't know what this `_adjusted` refers to but that's the only relevant difference between subject01.osim and subject01_adjusted.osim (the models also have different locations for the moving path points of the knee extensors but moving path point locations are actually not unused). A non-zero z-component of the torso mass center seems weird to me so I would tend to trust subject01.osim more than subject01_adjusted.osim, what do you think? Note that in the gait2392 folder, there is also subject01_simbody_adjusted that has the same torso mass center as subject01_adjusted but different inertia in general. I am not sure what the `_simbody` refers to but the inertia of the different segments seem pretty similar to those of Gait2354_Simbody/subject01_simbody.osim.

This is a little confusing but would that make sense to use [Gait2392_Simbody/subject01.osim](https://github.com/opensim-org/opensim-models/blob/master/Models/Gait2392_Simbody/subject01.osim#L2247) as the basis for the scaled models in the Gait10dof18musc folder? Would you be able to run the same diff analysis as you did before but using subject01.osim rather than subject01_adjusted.osim?

Let me know if this is unclear, thanks!

carmichaelong commented:
>  A non-zero z-component of the torso mass center seems weird to me so I would tend to trust subject01.osim more than subject01_adjusted.osim, what do you think?

:+1:

> would that make sense to use Gait2392_Simbody/subject01.osim as the basis for the scaled models in the Gait10dof18musc folder?

From looking at the files, I think this does make sense. Seems like the Gait10dof18musc subject01 model seems to be based on the Gait2392 subject01 model.

> Would you be able to run the same diff analysis as you did before but using subject01.osim rather than subject01_adjusted.osim?

Happy to do another diff, though I didn't quite follow which two files you want to compare and I want to double check. Is this comparing 1) gait2392's subject01.osim with 2) gait1018's subject01.osim?

antoinefalisse commented:
> Happy to do another diff, though I didn't quite follow which two files you want to compare and I want to double check. Is this comparing 1) gait2392's subject01.osim with 2) gait1018's subject01.osim?

Yes, that's what I meant! If the segment properties match, then I guess we can update the muscle properties of the gait1018's models by those of the gait2392's model.

carmichaelong commented:
Here's a summary of differences between gait2392's subject01 and gait1018's subject01. 

- There are many small differences (< 10^-5), which I have omitted from any comments here.
- For comparison with above, pelvis skeletal properties looks the same between these two:
![image](https://user-images.githubusercontent.com/7455790/96760186-b69d2300-138d-11eb-9c1d-18ab09974fa9.png)
- Gait 1018's glut_max muscle: doesn't seem to line up with any of gait2392's glut_max 3 muscles (w.r.t optimal fiber length, tendon slack length, path points). It's also hard to diff these sections by eye, so it's also possible I may have missed which one is lines up with.
- Quads moving path point differences:
![image](https://user-images.githubusercontent.com/7455790/96765869-1e07a280-138f-11eb-9332-73fa1c9cd6c9.png)
![image](https://user-images.githubusercontent.com/7455790/96765923-2e1f8200-138f-11eb-8a58-0065a91a36ea.png)
- Vas int muscle property differences (isometric force increase makes sense):
![image](https://user-images.githubusercontent.com/7455790/96766029-50b19b00-138f-11eb-86ac-f19c100dcfed.png)
- Medial gas muscle property differences (isometric force increase makes sense)
![image](https://user-images.githubusercontent.com/7455790/96766112-7343b400-138f-11eb-87dd-f4aeec96ae7c.png)
- Differences seems to be the same on each side too

For reference, here are the two model files below. If you're curious about the diff, what I do is drop them into the same folder, and then use TortoiseGit to diff them:
[subject01_models.zip](https://github.com/opensim-org/opensim-models/files/5418033/subject01_models.zip)

antoinefalisse commented:
Cool, thanks! So it seems that there are fewer differences between the scaled versions than between the unscaled versions (https://github.com/opensim-org/opensim-models/issues/129). I will update the PR with changes to the scaled gait1018 scaled models to make sure they match the scaled gait2392 models.

antoinefalisse commented:
@carmichaelong I have updated the PR with changes in the parameters of the scaled versions of `gait10dof18musc`. Those parameters should now match those of `Gait2392_Simbody/subject01.osim`.

I verified that: 

- All models were symmetrical.

- 2D and 3D models had the same muscle-tendon parameters (except for Fmax).
   

-  2D and 3D models had the same lengths and moment arms for a set of random poses (except for the ankle muscles, since the ankle definition is different in 3D vs 2D).

When doing those checks, I noticed slight differences (<1mm) in the parameters of `vast_int_l` vs `vast_int_r` for the scaled versions of gait2392. I have adjusted the left side to match the right side. This is a very small change but the models should now be symmetrical.

I think the PR is ready for review. Many thanks.

carmichaelong commented:
Thanks for all the detailed work @antoinefalisse! I've taken one pass as it and it's looking good so far. I'll look to take another pass tomorrow just with so many numbers to check through.

Before merging, we should double check with others about what tutorials/tests might change/break with these new models. @aymanhab do you know how changes here will propagate through core and gui repos? Not necessarily looking for exact answers, but maybe just some general thoughts so that I have a feel for what steps might need to be taken after this PR is merged so we don't leave core or gui in a weird state for our CI.

aymanhab commented:
Thanks @carmichaelong and @antoinefalisse gait2354 and 2392 are used in tutorials, if the test cases pass then very likely the changes are not significant since the test cases should cover the tutorials though would be good to verify. The model gait10dof18musc is a reduction of gait2392 that's not used in any specific tutorial AFAIK. Leg6dof9 (not discussed here but a distant cousin), is used in the strength of simulation tutorial RRA/CMC.

carmichaelong commented on Models/Gait10dof18musc/gait10dof18musc.osim:
```diff
@@ -2652,56 +2607,41 @@
 								</ConditionalPathPoint>
 								<MovingPathPoint name="vas_int_l-P4">
-									<location> 0.0548266 0.025002 -0.00177934</location>
+									<location> 0.0617576 0.020984 -0.0014</location>
```
This differs for the right side in this model, and may be the source of the symmetry differences for this exact point in all the other models too. IIRC, though, this MovingPathPoint isn't really used or am I mixing that up with another PathPoint? May be good to clean this up either way.

antoinefalisse replied:
Thanks @carmichaelong. I am pretty sure locations are not used with moving path points, since the location is determined by the spline (in this case). Let's talk about that during dev meeting. Thanks for the review.

aymanhab replied:
Indeed location is unused for moving path points and can be ignored.

antoinefalisse replied:
@carmichaelong I have set the location of the moving path points to Vec3(0) in the gait1018, gait2392, gait2354, and Hamner models. I also set the constant value of translation3 of the gait1018 models to 0.0 instead of 0.02 (not sure where this weird number came from). I had a tab issue so I replaced all spaces by tabs in one file, hope I did not mess things up. Let me know if I should do something else.

carmichaelong replied:
Edits to the model look good. Let me see if I can do something about the spaces vs tabs before we merge. @antoinefalisse looks like you used 4 spaces per tab?

antoinefalisse replied:
Yeah, isn't that the standard?

carmichaelong replied:
Yup it's the standard (and what OpenSim uses) but wanted to double check that my counting wasn't off. Note that other projects will sometimes use just 2 or 3 for economy of space though. You can see this sometimes on GitHub if you open a file with both leading spaces and tabs, they won't "line up" correctly if you use 4 spaces (i.e., below).

![image](https://user-images.githubusercontent.com/7455790/101080576-cf3c5380-355d-11eb-8881-33da1e1d5d60.png)

antoinefalisse replied:
Yes I had seen that bad alignment, so that's why there is one last commit to correct for that.

carmichaelong reviewed (commented):
@antoinefalisse Finally got to give this another close look today. Still looks great. I only noted some differences in a vas_int MovingPathPoint between the left and right sides of all the models, however, I feel like we discussed before that this may not be used in this model? Could be good to chat about this at a dev meeting real quick.

carmichaelong commented:
@antoinefalisse I just made a push to your branch to do spaces to tabs. Let me know if you think that looks good, and then I think we're good to merge.

antoinefalisse commented:
Looks good to me.
